### PR TITLE
 Support trending links

### PR DESCRIPTION
**Pull request by nikclayton**
Expand the "trending" support to include trending links.

Main changes:

- `TrendingActivity` can now host multiple trending fragments in a viewpager, and swipe between them
- Implement a `TrendingLinks` suite (`Fragment`, `ViewModel`, `Adapter`, `Repository`, etc) to get and display the data
- Always show "Trending" on the left-nav menu
- Support adding trending links as a dedicated tab

### CODE REVIEW

**Comment by nikclayton** [comment]
This includes the changes in https://github.com/tuskyapp/Tusky/pull/3906, which are mechanical renames. This PR will be easier to review if you review that one first.

Some screenshots:

**When added as a tab**



**As part of the activity**

**Comment by nikclayton** [comment]
I'm adding trending posts in another PR.

**Comment by Lakoja** [comment]
(Must this be updated somehow: it still shows the changes from the merged rename PR.)

((But also that PR is not shown under "closed PRs"...))

**Comment by nikclayton** [comment]
> (Must this be updated somehow: it still shows the changes from the merged rename PR.)

Yes. Done.

> ((But also that PR is not shown under "closed PRs"...))

Weird. It shows for me.

Note that the list of PRs is default-sorted by "When was the PR opened?", so a PR that was opened several weeks back, but was closed an hour ago, could easily show up on page 2 or 3 of the PR page.

File: app/src/main/java/com/keylesspalace/tusky/MainActivity.kt [hunk 1]
```diff
@@ -175,37 +180,47 @@ class MainActivity : BottomSheetActivity(), ActionButtonActivity, HasAndroidInje
...
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
         val activeAccount = accountManager.activeAccount
             ?: return // will be redirected to LoginActivity by BaseActivity
 
         var showNotificationTab = false
-        if (intent != null) {
+
+        // check for savedInstanceState in order to not handle intent events more than once
```
Comment (Lakoja) [inline app/src/main/java/com/keylesspalace/tusky/MainActivity.kt:192]: It checks for intent state? (savedInstanceState is null)

File: app/src/main/java/com/keylesspalace/tusky/MainActivity.kt [hunk 2]
```diff
@@ -175,37 +180,47 @@ class MainActivity : BottomSheetActivity(), ActionButtonActivity, HasAndroidInje
...
+        if (intent != null && savedInstanceState == null) {
```
Comment (Lakoja) [inline app/src/main/java/com/keylesspalace/tusky/MainActivity.kt:193]: All of the changes in MainActivity and other classes seem unrelated to this PR?

- Intent handling
- Context menu

Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/MainActivity.kt:193]: I think you're looking at the wrong diff. Those are changes from https://github.com/tuskyapp/Tusky/pull/3732 which has already been merged in to `develop` (and `develop` has been merged in to this PR).

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    fun bind(link: TrendsLink, statusDisplayOptions: StatusDisplayOptions) {
+        this.link = link
+
+        // TODO: This is very similar to the code in StatusBaseViewHolder.setupCard().
+        // Can a "PreviewCardView" type be created to encapsulate all this?
+        binding.cardTitle.text = link.title
+
+        val description = if (link.description.isNotBlank()) {
```
Comment (Lakoja) [inline app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt:54]: Would this look nicer with a "when"?

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        } else {
+            null
+        }
+        description?.let { binding.cardDescription.text = it } ?: binding.cardDescription.hide()
+
+        binding.cardLink.text = link.url
+
+        val cardImageShape = ShapeAppearanceModel.Builder()
+
+        if (statusDisplayOptions.mediaPreviewEnabled && !link.image.isNullOrBlank()) {
```
Comment (Lakoja) [inline app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt:67]: What does all this do (comment?)?

It seems to adapt layout dynamically but to what end?

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            }
+        }
+
+        viewLifecycleOwner.lifecycleScope.launch {
+            viewModel.statusDisplayOptions.collectLatest {
+                adapter.statusDisplayOptions = it
+            }
+        }
+
+        (activity as? ActionButtonActivity)?.actionButton?.hide()
```
Comment (Lakoja) [inline app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt:161]: Can it every be an ActionButtonActivity here?

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    private fun onOpenLink(url: String) = requireContext().openLink(url)
+
+    override fun onResume() {
+        super.onResume()
+        val a11yManager =
+            ContextCompat.getSystemService(requireContext(), AccessibilityManager::class.java)
+
+        val wasEnabled = talkBackWasEnabled
```
Comment (Lakoja) [inline app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt:220]: This is only done for the trending fragments?

(Looks like a job for a base class - and then used elsewhere, too?)

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksRepository.kt [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ * You should have received a copy of the GNU General Public License along with Tusky; if not,
+ * see <http://www.gnu.org/licenses>.
+ */
+
+package com.keylesspalace.tusky.components.trending
+
+import com.keylesspalace.tusky.network.MastodonApi
+import javax.inject.Inject
+
+class TrendingLinksRepository @Inject constructor(
```
Comment (Lakoja) [inline app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksRepository.kt:23]: This looks small? :-)

What is the future (extension) of this?

File: app/src/main/java/com/keylesspalace/tusky/entity/TrendsLink.kt [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    val blurhash: String?
+}
+
+data class LinkHistory(
+    val day: String,
+    val accounts: Int,
+    val uses: Int
+)
+
+data class TrendsLink(
```
Comment (Lakoja) [inline app/src/main/java/com/keylesspalace/tusky/entity/TrendsLink.kt:62]: What is a trends link (or a trending link in other places)?

(Maybe small comment.)

File: app/src/main/res/layout/item_trending_link.xml [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:paddingStart="14dp"
+    android:paddingEnd="14dp">
+
+    <LinearLayout
+        android:id="@+id/status_card_view"
```
Comment (Lakoja) [inline app/src/main/res/layout/item_trending_link.xml:27]: trending_link_card_view?

(Can an actual card be used?)

File: app/src/main/java/com/keylesspalace/tusky/TabData.kt [hunk 10]
```diff
@@ -98,6 +100,12 @@ fun createTabDataFromId(id: String, arguments: List<String> = emptyList()): TabD
             icon = R.drawable.ic_trending_up_24px,
             fragment = { TrendingTagsFragment.newInstance() }
         )
+        TRENDING_LINKS -> TabData(
```
Comment (Lakoja) [inline app/src/main/java/com/keylesspalace/tusky/TabData.kt:103]: It would make sense to only have that "trending activity view with tabs" you'll get from the side drawer.

So the (only) tab here would show that as well.

(Consistent and removes clutter from this tab list, this is especially relevant for the following PR when this gets three tabs here.)

Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/TabData.kt:103]: That would put a set of tabs (trending tags, links, posts) inside another set of tabs.

That's possible (Material3 calls them "secondary tabs") but it's a departure from how we've done things so far. I also think it's reasonable for a user to only want one or two of them and not all three (e.g., I'd use links and posts, but have no interest in trending hashtags).

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt (hunk 4 above)
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt:67]: There are two preview layout types.

One puts the preview image above the link title and description (e.g., see the screenshots in the second comment on this PR).

The other puts the preview to the left of the title and description. As the previous comment notes, this is out of StatusBaseViewHolder, and might be moved somewhere common later if it needs reusing again.

This is to follow what I call the rule-of-three for refactoring (see also https://en.wikipedia.org/wiki/Don%27t_repeat_yourself#AHA).

1. You don't need to refactor code that's only used once into a common function, because you don't have the necessary knowledge to know what parts of the code should be governed by parameters (you have zero examples of reuse).
2. You don't need to refactor code that's only used twice in to a common function, because you still don't have the necessary knowledge to know what parts of the code should be governed by parameters (you have one example of reuse, which is not enough).
3. When you've copy pasted the same code in to the three different places... now is the time to start looking at refactoring in to common function, because you've three distinct use cases that you can learn from (you have two examples of reuse, which is starting to become useful).

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksRepository.kt (hunk 7 above)
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksRepository.kt:23]: Yes, but it serves an important role -- it decouples the view model from the API.

The goal is:

- Fragments/Activities -- display the UI state, convert user actions in to actions to send to the view model
- ViewModel -- fetches data from repositories, converts it to UI state for display. Turns user actions from the UI into operations the repository should perform on the data
- Repository -- manages access to the data. Shields the ViewModel from how that data is stored / where it is fetched from
- PagingSource / RemoteMediator -- fetches the underlying data

https://developer.android.com/topic/architecture is the place to read about all this.

We're currently in the middle of transitioning the existing code to this. New code should follow this model.

File: app/src/main/java/com/keylesspalace/tusky/components/trending/viewmodel/TrendingLinksViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    private val _loadState = MutableStateFlow<LoadState>(LoadState.Initial)
+    val loadState = _loadState.asStateFlow()
+
+    private val _statusDisplayOptions = MutableStateFlow(
+        StatusDisplayOptions.from(preferences, account)
+    )
+    val statusDisplayOptions = _statusDisplayOptions.asStateFlow()
+
+//    val data = MutableStateFlow<List<TrendsLink>>(emptyList())
+
```
Comment (nikclayton) [inline app/src/main/java/com/keylesspalace/tusky/components/trending/viewmodel/TrendingLinksViewModel.kt:77]:
```suggestion
```

File: app/src/main/java/com/keylesspalace/tusky/entity/TrendsLink.kt (hunk 8 above)
Comment (nikclayton) [inline app/src/main/java/com/keylesspalace/tusky/entity/TrendsLink.kt:62]:
```suggestion
/** Represents a https://docs.joinmastodon.org/entities/PreviewCard/#trends-link */
data class TrendsLink(
```

File: app/src/main/res/layout/item_trending_link.xml (hunk 9 above)
Comment (nikclayton) [reply app/src/main/res/layout/item_trending_link.xml:27]: This is to keep it the same as the layout it was copied from. If these do get consolidated in to a dedicated view later it'll make doing the consolidation easier.

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt (hunk 3 above)
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt:54]: Sure, I'll do that.

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt (hunk 5 above)
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt:161]: Yes. `MainActivity`, which can host this as a tab, is an `ActionButtonActivity`.

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt (hunk 6 above)
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt:220]: No, it's done in `TimelineFragment` as well, so anything that inherits from that gets it too.

Before you ask, this can't inherit from `TimelineFragment`, because that's written to assume that it's showing a timeline of statuses, not a timeline of arbitrary other items.

File: app/src/main/java/com/keylesspalace/tusky/entity/TrendsLink.kt (hunk 8 above)
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/entity/TrendsLink.kt:63]: Done.

**Comment by nikclayton** [comment]
@Lakoja Please continue to ask questions, but please mark the ones that I've answered and you've understood as "Resolved", so it's easy for me to see whatever's new. Thanks.

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt (hunk 4 above)
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt:64]: One way to see the other layout is to disable downloading media, then it looks like this:

**Comment by Lakoja** [comment]
> @Lakoja Please continue to ask questions, but please mark the ones that I've answered and you've understood as "Resolved", so it's easy for me to see whatever's new. Thanks.

Unfortunately, I cannot do resolve them (or reopen, non-contributor?). Otherwise I would've done it..

File: app/src/main/java/com/keylesspalace/tusky/TabData.kt (hunk 10 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/TabData.kt:103]: Ok...
I see (secondary tabs).

But it would still be (quite) inconsistent:

- One location with a combined view (from side drawer)
- And one location where I would need to make a choice; and/or add 3 things seperately to get the same thing I get from the side drawer

:beard_scratch:

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt (hunk 4 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinkViewHolder.kt:64]: > There are two preview layout types.
> 
> One puts the preview image above the link title and description (e.g., see the screenshots in the second comment on this PR).
> 
> The other puts the preview to the left of the title and description. As the previous comment notes, this is out of StatusBaseViewHolder, and might be moved somewhere common later if it needs reusing again.

These might be the code comment(s) I was looking for.

File: app/src/main/java/com/keylesspalace/tusky/entity/TrendsLink.kt (hunk 8 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/entity/TrendsLink.kt:63]: It says there that it is a "trends link" - but no further explanation.

**Comment by Lakoja** [comment]
> > > @Lakoja Please continue to ask questions, but please mark the ones that I've answered and you've understood as "Resolved", so it's easy for me to see whatever's new. Thanks.
> > 
> > 
> > Unfortunately, I cannot do resolve them (or reopen, non-contributor?). Otherwise I would've done it..
> 
> Huh. Today I learned, I guess. OK, please just add a comment to any of the ones that can be resolved.

Yeah. I was wondering about that.
At least my own i would like to resolve or open...

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt (hunk 5 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt:161]: (can be resolved)

File: app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt (hunk 6 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingLinksFragment.kt:220]: (can be resolved)

File: app/src/main/res/layout/item_trending_link.xml (hunk 9 above)
Comment (Lakoja) [reply app/src/main/res/layout/item_trending_link.xml:27]: (can be resolved)

**Comment by nikclayton** [comment]
> Yeah. I was wondering about that. At least my own i would like to resolve or open...

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#resolving-conversations says you should be able to resolve comments on the PRs that you have opened.

File: app/src/main/java/com/keylesspalace/tusky/TabData.kt (hunk 10 above)
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/TabData.kt:103]: @Lakoja Yes, so let's think about two typical users.

User A occasionally checks the trending info, but not habitually. Maybe a few times a week to see if there's anything new. They don't have a need to focus on one of the three (soon to be four) types of thing that can trend, so getting to it from the nav menu when they want to, and tapping through the different tabs to see what's new is fine for them.

User B is much more interested in trending links, they want to check regularly as part of their normal Mastodon experience. So much so that they want to make it a tab. They're not interested in posts or hashtags, just links. For them, adding a tab with just trending links is fine.

If they added a tab that contained all the trending items they would have to tap that tab once to choose "trending", and then tap again to choose which trending item they're interested in. That's suboptimal UX for an activity that User B is going to want to do a lot.

That's why I think the UX in this PR is the way to go.

There's a third user, User C, who are like User B, but they want to check all the trending tabs. They're not well served at the moment, because of the 5-tab limit, but I think that should be removed (see discussion in "Tusky General" channel).
